source/lac/trilinos_tpetra_sparsity_pattern.cc

```diff
@@ -0,0 +1,1078 @@
+             ExcInternalError());
+    }
+
+
+
+    template <typename Node>
+    bool
+    SparsityPattern<Node>::row_is_stored_locally(const size_type i) const
+    {
+      return graph->getRowMap()->getLocalElement(i) !=
+             Teuchos::OrdinalTraits<int>::invalid();
+    }
+
+
+
+    template <typename Node>
+    bool
+    SparsityPattern<Node>::exists(const size_type i, const size_type j) const
+    {
+      if (!row_is_stored_locally(i))
+        return false;
+
+      // Extract local indices in  the matrix.
+      auto trilinos_i = graph->getRowMap()->getLocalElement(i);
+      auto trilinos_j = graph->getColMap()->getLocalElement(j);
+
+      typename GraphType::local_inds_host_view_type col_indices;
+
+      // Generate the view.
+      graph->getLocalRowView(trilinos_i, col_indices);
+
+      // Search the index
+      const size_type local_col_index =
+        std::find(col_indices.data(),
+                  col_indices.data() + col_indices.size(),
+                  trilinos_j) -
+        col_indices.data();
+
+      return static_cast<size_t>(local_col_index) != col_indices.size();
+    }
+
+
+
+    template <typename Node>
+    typename SparsityPattern<Node>::size_type
+    SparsityPattern<Node>::bandwidth() const
+    {
+      size_type                         local_b  = 0;
+      TrilinosWrappers::types::int_type global_b = 0;
+      for (int i = 0; i < static_cast<int>(local_size()); ++i)
+        {
+          typename GraphType::local_inds_host_view_type indices;
+          graph->getLocalRowView(i, indices);
+          int num_entries = indices.size();
+          for (unsigned int j = 0; j < static_cast<unsigned int>(num_entries);
+               ++j)
+            {
+              if (static_cast<size_type>(std::abs(i - indices[j])) > local_b)
+                local_b = std::abs(i - indices[j]);
+            }
+        }
+
+      global_b =
+        Utilities::MPI::max(local_b,
+                            Utilities::Trilinos::teuchos_comm_to_mpi_comm(
+                              graph->getComm()));
+      return static_cast<size_type>(global_b);
+    }
+
+
+
+    template <typename Node>
+    unsigned int
+    SparsityPattern<Node>::local_size() const
+    {
+      int n_rows = graph->getLocalNumRows();
+
+      return n_rows;
+    }
+
+
+
+    template <typename Node>
+    std::pair<typename SparsityPattern<Node>::size_type,
+              typename SparsityPattern<Node>::size_type>
+    SparsityPattern<Node>::local_range() const
+    {
+      size_type begin, end;
+      begin = graph->getRowMap()->getMinGlobalIndex();
+      end   = graph->getRowMap()->getMaxGlobalIndex() + 1;
+
+      return std::make_pair(begin, end);
+    }
+
+
+
+    template <typename Node>
+    std::uint64_t
+    SparsityPattern<Node>::n_nonzero_elements() const
+    {
+      TrilinosWrappers::types::int64_type nnz = graph->getGlobalNumEntries();
+
+      return static_cast<std::uint64_t>(nnz);
```

Comment:
Let's let the compiler figure out which casts are necessary:
```suggestion
      return graph->getGlobalNumEntries();
```